File: aoc-2025/day03.py
with open("day03.in") as f:
    inp = [[int(c) for c in s] for s in f.read().strip().splitlines()]

def part1(row):
    cur = 0
    for i in range(len(row)):
        for j in range(i + 1, len(row)):
            cur = max(cur, 10 * row[i] + row[j])
    return cur

print(sum(map(part1, inp)))

def part2(row):
    deletions = len(row) - 12
    stack = []

    for n in row:
        while stack and deletions > 0 and n > stack[-1]:
            stack.pop()
            deletions -= 1
        stack.append(n)

    return (int("".join(map(str, stack[:12]))))

print(sum(map(part2, inp)))

File: aoc-2025/day03.ml
let input =
  let parse_row s =
    s
    |> String.to_seq
    |> List.of_seq
    |> List.map (fun c -> int_of_char c - int_of_char '0')
  in
  open_in "day03.in"
  |> In_channel.input_all
  |> String.trim
  |> String.split_on_char '\n'
  |> List.map parse_row

let largest_sub target row =
  let rec aux stack deletions = function
    | [] -> List.rev stack
    | hd :: tl ->
        let rec pop s d =
          match s with
          | hd' :: tl' when d > 0 && hd > hd' -> pop tl' (d - 1)
          | _ -> s, d
        in
        let stack, remaining_deletions = pop stack deletions in
        aux (hd :: stack) remaining_deletions tl
  in
  row
  |> aux [] (List.length row - target)
  |> List.take target
  |> List.fold_left (fun acc d -> acc * 10 + d) 0 

let solve n = List.fold_left ( + ) 0 (List.map (largest_sub n) input)
let part1 = solve 2
let part2 = solve 12

let () = Printf.printf "Part 1 = %d, Part 2 = %d\n" part1 part2
